### ISSUE TITLE
Any news when this is expected to be ready for production 

### ISSUE THREAD

**Issue by almogdesign** [CollectionFS/Meteor-CollectionFS#534]
Any news when this is expected to be ready for production

**Comment by aldeed** [comment]
Depends on the level of production. Small-scale production might be OK now, though I would stop short of recommending it since there are still a dozen or so known issues. There are also probably 20-30 features that should be added or refactored before it can be considered production caliber. And we need to iron out the worker process and direct-to-cloud SAs before we can call it truly scalable.

If it continues to be mostly @raix and I doing the development in very little spare time, then at least 6 months. If we can rally a bit more community involvement, then could be much less.

**Comment by theduke** [comment]
I would like to chip in and help out a bit. 
I just need some pointers where to start.

**Comment by wursttheke** [comment]
I'd also like to help out, but it seems to be quite a vast project. Dozens of packages in different reps. Absolutely unsure where to start. :)

Is there a list of issues / features that need some love? Maybe a Trello board or something?

**Comment by raix** [comment]
Hi guys,

Would be great with some help @aldeed and I will try answering as best possible - we are both very booked atm. but I've reserved some time next week (my vacation actually) simply to get everything up and running - not sure how many issues I get closed. We could have a developers hangout next week and talk about things?

I think its a good idea to have a trello board to get an overview of things - I did a diagram a long time ago in word or excel - just a really basic overview of the project for me an Eric to work out from.

So in short the main thing is to write tests for the base packages - we cant really depend on cfs until tests are written and pass - its also a great way of learning about the system.

We have some issues about partial download support and mobile support. The first is about actually implementing the partial support in the SA's eg. gridFS etc. The second is about ensuring UInt8Arrays etc. work so its prop. a good idea to write unit testing for uploading.

I'll just add a trello board in a sec

**Comment by aldeed** [comment]
Dev hangout next week sounds good. There is theoretically a doc with extra info for developers [here](https://github.com/CollectionFS/Meteor-CollectionFS/blob/devel/ADVANCED.md), but parts of it need updating. I can try to brain dump some architecture stuff there and maybe we can get @raix's diagram updated, too.

I've been slowly trying to get issues closed on the component package repos and move all remaining to this repo so that we have a single work list.

There are a lot of github issues, but many are probably duplicates or symptoms of the same underlying issue and many others can probably be closed. We've just been really far behind on tagging and closing them.

**Comment by raix** [comment]
@aldeed We could merge all the package repo's into one main repo eg. having all the packages as subfolders - it would help centralize issues and it would give a better overview of the project?
And we would have a more "track" feeling about the project concerning versions etc.

**Comment by raix** [comment]
+ we could rig a bash version / publish script - making sure all packages are aligned and published.

**Comment by raix** [comment]
Trello: https://trello.com/b/iFsuP745/meteor-collectionfs

**Comment by raix** [comment]
We should prop. do a shoutout on Meteor talk+core about this, making sure we catch all interested in the project?

**Comment by aldeed** [comment]
The only issue that might come up with merging into one repo is we can't give developers access to only certain packages. I don't know if that's something we even need to do, but it's the one drawback I can think of.

**Comment by raix** [comment]
@aldeed should we close repo's for issues letting only the main repo have issues?

**Comment by aldeed** [comment]
That's a good idea.

**Comment by wursttheke** [comment]
Thanks @raix for adding me to the Trello board. I'm not sure how much help I can be here, though. Looking at your overview file, I'm actually quite overwhelmed with all the moving parts of this system. 

I'm not really in the position to have an opinion here, but I would suggest to reduce the overall complexity as much as possible and kind of "throw out" everything not really needed for a stable basic package v1.0.0 which can be used with confidence. To me, a lot of features seem to be still in the ideas stage or are only partly implemented. E.g the direct S3 upload feature is nice, but something I would move to a future-ideas list. 

I would also reduce the number of repos quite a bit for exactly the same reason: not to overwhelm new developers and make it easy to contribute. :)

**Comment by rhyslbw** [comment]
+1 on the Trello board and hangout. @aldeed and @raix have done an amazing job but this project needs a bigger team, so glad to see this.

Considering some of the functionality has been built as a package that can be used as a standalone component elsewhere, it's not going to be possible to consolidate everything under the one repo, but certainly close off CFS-only package repo issues. I'd suggest managing a 1.0.0 task list and effort in Trello with a dedicated documentation site to bring it all together for implementers or developers. This would then remove the necessity to structure the Github org in a more digestable way for someone fresh.

**Comment by theduke** [comment]
Could I get an invite too? Email: reg@theduke.at.
---------

Regarding the project structure: I used to split up larger projects into many small sub-repositories too, and it always ended up being a huge development overhead and big cause of confusion for both new and existing project members. I would strongly vote for consolidating everything into one well-structured repository! After all, the whole Linux kernel is in one repo as well!

Of course, it would be an unwelcome overhead to force users of the package to load all the code on the client and server, so it would be good to have several Meteor packages defined in the big repo, which can still be published separately to Atmosphere, so devs can still only use the parts that they need.

The way it is now it's just hard to dive in.

**Comment by raix** [comment]
@theduke added
I've put "How can we make it easier to contribute" on the agenda, I think its important to us all that cfs keeps a fair level of maintenance - Eric and I can't keep up at the moment + more heads = better ideas

@wursttheke the diagram is a bit outdated but shows a general direction - the idea about splitting everything up into smaller modules makes it easier to test etc. problem is tracking issues and keeping an overview of related issues across repo's.

I actually think the packages should be even smaller building bricks, it makes it easier to reason about/write unit tests and makes it more reusable - that said we should prop. consider having the core packages in one repo, but we can talk more about it on wed.

We have milestones on the main repo, and stuff like direct upload etc. is on there, but its not for v1

I'm creating a separate issue inviting more devs

**Comment by gabrielstuff** [comment]
The project seems to get more and more traction. I'll be happy to help if I knew where to start.

**Comment by raix** [comment]
Hi @almogdesign I'm just linking some more info https://github.com/CollectionFS/Meteor-CollectionFS/issues/539#issuecomment-75376195

I'm refactoring / merging the core packages into one repo - this will help us all to get a better overview - https://trello.com/c/MT3P4hmv

**Comment by raix** [comment]
Sorry @almogdesign I meant @gabrielstuff ...

**Comment by gabrielstuff** [comment]
@raix, do you have any link to contribution ?I'm trying to start contributing but unless creating new packages I've never patch / contribut to existing ones.
Thanks

**Comment by raix** [comment]
The basics is to clone, create a new branch from the devel branch (if its not found we should create one, we should not use master), then submit a pr. We'll label it "in review" and do comments etc. Then merge and test/do qa - At release points we merge devel to master and publish.

Note we have previously committed directly to master, but we all have to do pr's from now on and have others review. (This should be documented...)

I'm in the middle of merging core packages mentioned in https://trello.com/c/sEgJ806r - so you can't contribute to those just yet (should be ready within hours)

Let me know where you want to start @gabrielstuff we'll answer questions and try to our best supporting :)

cc. @aldeed (regarding the pr / review procedure)

**Comment by gabrielstuff** [comment]
ok thanks @raix! that is what I've done for patching issue about video (#500 and #502). What you guys thought about a gitter room ?

**Comment by raix** [comment]
@gabrielstuff gitter badge added

**Comment by aldeed** [comment]
Currently `master` in the main repo is actually the old CFS v1 and we've been committing directly to `devel`. For all the other packages, we've been committing directly to `master`. After we have merged all packages into the main repo, I agree that we should do feature branches off `devel` and PRs to merge them back into `devel`. Then occasional version bump, publish, and merge into `master`.

**Comment by raix** [comment]
Lets clean up devel a bit, then set it as master (Master should be lastest published version)
